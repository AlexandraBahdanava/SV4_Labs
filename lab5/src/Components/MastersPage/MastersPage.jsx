import React, { useState, useEffect } from 'react';
import MastersCard from '../Card/MastersCard';
import CardForm from '../CardForm/CardForm';
import { Container, Grid, Button, Typography } from '@mui/material';
import SearchBar from '../SearchBar/SearchBar';
import ExperienceSlider from '../ExperienceSlider/ExperienceSlider';
import SpecializationDropdown from '../SpecializationDropdown/SpecializationDropdown';

function MastersPage() {
    const [masters, setMasters] = useState([]);
  const [selectedIds, setSelectedIds] = useState(new Set());
  const [isFormOpen, setIsFormOpen] = useState(false);
  const [editingMaster, setEditingMaster] = useState(null);
  const [filteredMasters, setFilteredMasters] = useState([]);
  const [isSearchPerformed, setIsSearchPerformed] = useState(false);
  const [experienceFilter, setExperienceFilter] = useState([0, 10]);
  const [selectedSpecialization, setSelectedSpecialization] = useState('');

    useEffect(() => {
        fetch('/data/masterData.json')
            .then(response => response.json())
            .then(data => setMasters(data))
            .catch(error => console.error("Ошибка при загрузке мастеров:", error));
    }, []);

    const handleSpecializationChange = (value) => {
        setSelectedSpecialization(value);
    };
  
    const handleSearch = (query) => {
        const filteredMasters = masters.filter(master =>
            master.name.toLowerCase().includes(query.toLowerCase()) &&
            (selectedSpecialization === '' || master.features.some(feature => feature.toLowerCase() === selectedSpecialization.toLowerCase())) &&
            master.experience >= experienceFilter[0] && master.experience <= experienceFilter[1]
        );
        setFilteredMasters(filteredMasters);
        setIsSearchPerformed(true);
    };

    const handleApplySpecializationFilter = () => {
        // Выполните здесь обновление результатов поиска, учитывая выбранную специализацию
        handleSearch(''); // Передайте пустой запрос, чтобы снова выполнить поиск с учетом специализации
    };

    const handleExperienceFilterChange = (value) => {
        setExperienceFilter(value);
        filterMastersByExperience(value);
      };

    const filterMastersByExperience = (value) => {
        const filteredMasters = masters.filter(master =>
          master.experience >= value[0] && master.experience <= value[1]
        );
        setFilteredMasters(filteredMasters);
      };

    const addOrUpdateMaster = (masterData, id) => {
        if (id) {
            setMasters(prevMasters =>
                prevMasters.map(master => master.id === id ? { ...master, ...masterData } : master)
            );
        } else {
            setMasters(prevMasters => [...prevMasters, { ...masterData, id: Date.now() }]);
        }
        setIsFormOpen(false);
        setEditingMaster(null);
    };

    const handleAddClick = () => {
        setEditingMaster(null);
        setIsFormOpen(true);
    };

    const handleEditClick = () => {
        const selectedIdArray = Array.from(selectedIds);
        if (selectedIdArray.length === 1) {
            const masterToEdit = masters.find(master => master.id === selectedIdArray[0]);
            setEditingMaster(masterToEdit);
            setIsFormOpen(true);
        } else {
            alert("Выберите объект для редактирования");
        }
    };

    const toggleSelect = (id) => {
        setSelectedIds(prevSelectedIds => {
            const newSelectedIds = new Set(prevSelectedIds);
            if (newSelectedIds.has(id)) {
                newSelectedIds.delete(id);
            } else {
                newSelectedIds.add(id);
            }
            return newSelectedIds;
        });
    };

    const removeSelectedMasters = () => {
        setMasters(prevMasters => prevMasters.filter(master => !selectedIds.has(master.id)));
        setSelectedIds(new Set());
    };

    return (
        <Container maxWidth="lg">
            <Grid container spacing={2} style={{ marginBottom: '20px', justifyContent: 'center' }}>
                <Grid item>
                    <Button variant="contained" style={{ backgroundColor: 'rgb(187, 87, 118)' }} onClick={handleAddClick}>
                        Добавить
                    </Button>
                </Grid>
                <Grid item>
                    <Button style={{ backgroundColor: 'rgb(187, 87, 118)' }} variant="contained" onClick={handleEditClick} disabled={selectedIds.size !== 1}>
                        Редактировать
                    </Button>
                </Grid>
                <Grid item>
                    <Button style={{ backgroundColor: 'rgb(187, 87, 118)' }} variant="contained" onClick={removeSelectedMasters}>
                        Удалить
                    </Button>
                </Grid>
            </Grid>

            <SearchBar onSearch={handleSearch} />
            <ExperienceSlider onFilterChange={handleExperienceFilterChange} />
             <SpecializationDropdown
                selectedSpecialization={selectedSpecialization}
                onSpecializationChange={handleSpecializationChange}
                onApplyClick={handleApplySpecializationFilter}
            />

            {isSearchPerformed ? (
                filteredMasters.length === 0 ? (
                    <Typography variant="body1" align="center" color="textSecondary">
                        Нет мастеров, соответствующих вашему запросу.
                    </Typography>
                ) : (
                    <Grid container spacing={2}>
                        {filteredMasters.map(master => (
                            <Grid item xs={12} sm={6} md={4} lg={3} key={master.id}>
                                <MastersCard
                                    master={master}
                                    isSelected={selectedIds.has(master.id)}
                                    toggleSelect={() => toggleSelect(master.id)}
                                />
                            </Grid>
                        ))}
                    </Grid>
                )
            ) : (
                <Grid container spacing={2}>
                    {masters.map(master => (
                        <Grid item xs={12} sm={6} md={4} lg={3} key={master.id}>
                            <MastersCard
                                master={master}
                                isSelected={selectedIds.has(master.id)}
                                toggleSelect={() => toggleSelect(master.id)}
                            />
                        </Grid>
                    ))}
                </Grid>
            )}

            {isFormOpen && (
                <CardForm
                    onSave={addOrUpdateMaster}
                    existingMaster={editingMaster}
                    isFormOpen={isFormOpen}
                    setIsFormOpen={setIsFormOpen}
                />
            )}
        </Container>
    );
}

export default MastersPage;
