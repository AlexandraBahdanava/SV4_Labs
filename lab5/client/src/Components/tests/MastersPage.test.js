import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import MastersPage from '../MastersPage'; // Подставьте правильный путь к вашему компоненту

describe('MastersPage', () => {
  it('renders correctly', () => {
    const { getByText } = render(<MastersPage />);
    expect(getByText('Поиск')).toBeTruthy();
    // Подставьте другие проверки, если необходимо
  });

  it('matches the snapshot', () => {
    const tree = renderer.create(<MastersPage />).toJSON();
    expect(tree).toMatchSnapshot();
  });

 
  // Добавьте другие тесты, если необходимо
});
