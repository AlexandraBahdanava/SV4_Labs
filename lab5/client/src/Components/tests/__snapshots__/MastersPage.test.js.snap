// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MastersPage matches the snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
>
  <div
    className="MuiBox-root css-mbx0dj"
  >
    <div
      className="MuiFormControl-root MuiTextField-root box css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id=":r0:"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Введите имя мастера"
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-ihdtdm"
          >
            <span
              className="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium buttonSearch css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Поиск
    </button>
  </div>
  <div>
    <p
      className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
      id="experience-slider"
    >
      Опыт работы (нажмите кнопку "Поиск" и начните перемещать ползунок):
    </p>
    <span
      className="MuiSlider-root MuiSlider-colorPrimary MuiSlider-sizeMedium css-187mznn-MuiSlider-root"
      onMouseDown={[Function]}
    >
      <span
        className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
      />
      <span
        className="MuiSlider-track css-1gv0vcd-MuiSlider-track"
        style={
          Object {
            "left": "0%",
            "width": "100%",
          }
        }
      />
      <span
        className="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-jx1sa5-MuiSlider-thumb"
        data-index={0}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "left": "0%",
            "pointerEvents": undefined,
          }
        }
      >
        <input
          aria-labelledby="experience-slider"
          aria-orientation="horizontal"
          aria-valuemax={10}
          aria-valuemin={0}
          aria-valuenow={0}
          data-index={0}
          disabled={false}
          max={10}
          min={0}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          step={1}
          style={
            Object {
              "border": 0,
              "clip": "rect(0 0 0 0)",
              "direction": "ltr",
              "height": "100%",
              "margin": -1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "whiteSpace": "nowrap",
              "width": "100%",
            }
          }
          type="range"
          value={0}
        />
        <span
          aria-hidden={true}
          className="MuiSlider-valueLabel css-7d61z1-MuiSlider-valueLabel"
        >
          <span
            className="MuiSlider-valueLabelCircle"
          >
            <span
              className="MuiSlider-valueLabelLabel"
            >
              0
            </span>
          </span>
        </span>
      </span>
      <span
        className="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-jx1sa5-MuiSlider-thumb"
        data-index={1}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "left": "100%",
            "pointerEvents": undefined,
          }
        }
      >
        <input
          aria-labelledby="experience-slider"
          aria-orientation="horizontal"
          aria-valuemax={10}
          aria-valuemin={0}
          aria-valuenow={10}
          data-index={1}
          disabled={false}
          max={10}
          min={0}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          step={1}
          style={
            Object {
              "border": 0,
              "clip": "rect(0 0 0 0)",
              "direction": "ltr",
              "height": "100%",
              "margin": -1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "whiteSpace": "nowrap",
              "width": "100%",
            }
          }
          type="range"
          value={10}
        />
        <span
          aria-hidden={true}
          className="MuiSlider-valueLabel css-7d61z1-MuiSlider-valueLabel"
        >
          <span
            className="MuiSlider-valueLabelCircle"
          >
            <span
              className="MuiSlider-valueLabelLabel"
            >
              10
            </span>
          </span>
        </span>
      </span>
    </span>
  </div>
  <div>
    <div
      className="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
      style={
        Object {
          "marginRight": 10,
          "minWidth": 200,
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        id="specialization-label"
      >
        Специализация
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        onClick={[Function]}
      >
        <div
          aria-controls=":r1:"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="specialization-label specialization-select"
          className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id="specialization-select"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="combobox"
          tabIndex={0}
        >
          <span
            className="notranslate"
          >
            ​
          </span>
        </div>
        <input
          aria-hidden={true}
          aria-invalid={false}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          disabled={false}
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value=""
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-yjsfm1"
          >
            <span>
              Специализация
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
  />
</div>
`;
